### PR TITLE
Respect policy list expiration

### PR DESCRIPTION
**Pull request by Snawoot**
This pull request contains changes discussed in [PR in old repo](https://github.com/EFForg/starttls-everywhere/pull/87).

Here is excerpt from old discussion about unresolved questions to new code.

## About mocking FS IO

> Instead of actually writing to policy.json, it may be cleaner to use mock.mock_open to mock all the open calls.

> What do you think about mocking the open methods rather than writing these files to /tmp?

`mock` is a great tool, but I doubt about it's purpose here. Mocking actual file IO allows us to run tests clearly, without interacting with real FS. But creation of temporary directory with file is not something we can't afford.

In general, best practice is to mock as less as possible, resort to mock only when counterpart our code tested against cannot be efficiently represented by real instance.

Mock solution has following benefit over real FS write: it avoids usage of real FS. But it has drawbacks:
* Reader will **always** read `read_data` with same type of string `mock_open` was initialized. Consider [following examples](https://gist.github.com/Snawoot/33b797bad8265a78f1071626d41de559). So, `mock_open` is imperfect.
* Reading of file by a program may produce different results depending on method of reading. For example, there are no reasons why `policy.json` can't have byte order mark in the beginning of file and to handle such case we have to use special BOM-aware codec. Mocked file defined by a string doesn't allow us to conduct tests against real files.
* Mocking requires test to have some knowledge of internal structure of program in order it can be properly patched, and this knowledge has to be maintained. What if we will use some faster json module which will load json file directly?

BTW, when our project will become popular (I believe!) and policy list will be big enough to not fit in memory, we probably will have to reconsider parsing of json file anyway in order to reduce memory footprint required to generate policy. Maybe generator-like json parser like `ijson` will do.

I think if we will need to mock file IO, we can return to it in future. But for now it is unnecessary.

## About tests generation

> Generating these tests dynamically is nice, but could be opaque-- for instance, if a particular test fails here, it may be difficult for someone unfamiliar with the project to trace back where the failure occurred.
> 
> What do you think about having a single test_generate that instead loops over the test cases and performs each one?

If I make some test fail (for example, by modifying expected result) it will clearly indicate which test has failed, like this:
```
...
=================================== FAILURES ===================================
___________________ TestPostfixGenerator.test_expired_policy ___________________
...
```

All testcases in test matrix have ID (first element in tuple). This ID is used to generate test name when it is added to `TestPostfixGenerator`. Its commonly established practice and below I'll provide some details about reasons.

I've considered many options when I was seeking for solution. Most obvious way and common practice is to use `@pytest.mark.parametrize` decorator. This decorator accepts matrix for test cases which optionally marked by ID used to infer test name. But I rejected this solution because:
* Current tests doesn't depend on pytest (but can be run with it). So I had to stick to `unittest` API in order to retain current independence from external testing framework.
* Single method `unittest.TestCase` cannot be parameterized because it can accept only one parameter with class instance by definition of its interface.

Considering ways which are compatible with `unittest` there are few options:
1. Use external package `parameterized` which provides decorator doing essentially same thing I do now, but adds external dependency.
2. Simply add test methods linked to testcases via closure in simple loop. (Actual implementation)
3. Use metaclasses and other complicated mechanisms in order to do job similar to option 2.

Solution with single test function iterating over config matrix is tempting, but I didn't chose it because it is not transparent to test runner:
1. All configs are collected by test runner as a single test. In my implementation it is distinct named tests.
2. For this reason it is not possible to enumerate tests properly (like `pytest --collect-only`) or run specific config test without running all other config tests. My implementation allows it. Example: `pytest tests/configure_test.py::TestPostfixGenerator::test_simple_policy`
3. External tools for building automated test reports (like Allure) will not be able to produce proper output about found defects.

In other words, this solution is somewhat simpliest way to parameterize tests, preserving proper testcase semantics and compatibility with built-in `unittest` library.

### CODE REVIEW

**Comment by sydneyli** [comment]
Thanks for your thorough comments, @Snawoot!
I'm okay with not mocking.

Regarding testing:
> Most obvious way and common practice is to use @pytest.mark.parametrize decorator.

> Use external package `parameterized` which provides decorator doing essentially same thing I do now, but adds external dependency.

Is there a significant cost of adding dev dependencies to the project? I expect this won't be the only place we might need test parameterization.

I've used [unittest.subTest](https://docs.python.org/3/library/unittest.html#unittest.TestCase.subTest) before for unittest parameterization, but to maintain compatibility with python 2 we'd have to use an external backports package like [unittest2](https://pypi.org/project/unittest2/). I might prefer this option over rolling our own parameterization, since it's (somewhat) native Python.

**Comment by Snawoot** [comment]
> Is there a significant cost of adding dev dependencies to the project?

No, my concern is not about size of dev dependencies. Project tests are virtually versatile and I think its good. One may pick any `unittest.TestCase`-aware framework: `unittest`, `pytest`, `nose` - and it will run current tests without a hitch. But if we will pick `unittest.subTest` - pytest [will not work](https://docs.pytest.org/en/latest/unittest.html). If we will pick `@pytest.mark.parametrize` all other runners except pytest will not work.

> I expect this won't be the only place we might need test parameterization.

If you like, I can rewrite current test wrapper to decorator reusable across all parameterized tests. I guess 10-15 lines of self baked code is not a big price. In return we'll receive tests which run everywhere with fine-grained control and meaningful reporting.

File: starttls_policy/tests/configure_test.py [hunk 1]
```diff
@@ -86,5 +122,24 @@ def test_instruct_string(self):
         self.assertTrue("postfix reload" in instructions)
         self.assertTrue(generator.default_filename in instructions)
 
+def enclose_config_case(conf, expected):
```
Comment (sydneyli) [inline starttls_policy/tests/configure_test.py:125]: naming nit: create_config_test ?

File: starttls_policy/tests/configure_test.py [hunk 2]
```diff
@@ -86,5 +122,24 @@ def test_instruct_string(self):
...
+            pol_filename = os.path.join(testdir, generator.default_filename)
+            with open(pol_filename) as pol_file:
+                result = pol_file.read()
+            os.remove(pol_filename)
+        self.assertEqual(result, expected)
+    return test
+
+for tname, a, b in testgen_data:
+    test_name = "test_%s" % tname
+    wrapped_test = enclose_config_case(a, b)
```
Comment (sydneyli) [inline starttls_policy/tests/configure_test.py:141]: naming nit: prefer longer, detailed names: `config_name, config_content, expected_output in testgen_data`

File: starttls_policy/tests/configure_test.py [hunk 3]
```diff
@@ -86,5 +122,24 @@ def test_instruct_string(self):
         self.assertTrue("postfix reload" in instructions)
         self.assertTrue(generator.default_filename in instructions)
 
+def enclose_config_case(conf, expected):
+    """Wrap testcase against config into closure in order to make it attachable
+    to main testcase class"""
```
Comment (sydneyli) [inline starttls_policy/tests/configure_test.py:127]: Maybe: Returns parametrized test for transforming policy file into Postfix config

**Review by sydneyli — APPROVED** [review]
> But if we will pick `unittest.subTest` - pytest [will not work](https://docs.pytest.org/en/latest/unittest.html).

That makes me so sad! I'm okay with rolling our own because it's fairly simple, and I agree that the flexibility to choose different testing frameworks at this early stage is pretty nice.

I just have a couple of naming nits to make it clearer that we're just parametrizing tests here, but other than that LGTM. Once you've made the naming changes feel free to merge this in.

EDIT: you can name them differently if you'd like too, I'd just like the general spirit to be preserved (`create` as a verb instead of `enclose`, and anything to be slightly more descriptive than `a`/`b` is ok too)

**Comment by Snawoot** [comment]
I'll resolve this issues and come up with all fixed tests in new pull request. Also I'll fix dependency of abstract generator tests on symlink.
